File: src/main/scala/com/vc/medium/L364.scala
package com.vc.medium


// This is the interface that allows for creating nested lists.
// You should not implement it, or speculate about its implementation
class NestedInteger364 {

  // Return true if this NestedInteger holds a single integer, rather than a nested list.
  def isInteger: Boolean = true

  // Return the single integer that this NestedInteger holds, if it holds a single integer
  def getInteger: Int = 1

  // Set this NestedInteger to hold a single integer.
  def setInteger(i: Int): Unit = Unit

  // Return the nested list that this NestedInteger holds, if it holds a nested list
  def getList : List[NestedInteger364] = Nil

  // Set this NestedInteger to hold a nested list and adds a nested integer to it.
  def add(ni: NestedInteger364): Unit = Unit
}

object L364 {
  def depthSumInverse(nestedList: List[NestedInteger364]): Int = {
    import scala.collection.mutable
    val q = new mutable.Queue[NestedInteger364]()
    nestedList.foreach(x => q.enqueue(x))
    var prev = 0
    var total = 0
    while(q.nonEmpty) {
      val size = q.size
      var levelSum = 0
      (0 until size).foreach(_ => {
        val e = q.dequeue
        if(e.isInteger) levelSum += e.getInteger
        else e.getList.foreach(x => q.enqueue(x))
      })
      prev += levelSum
      total += prev
    }
    total
  }
}

File: src/main/java/com/vc/medium/L364.java
package com.vc.medium;

import com.vc.hard.NestedInteger;

import java.util.*;

class L364 {
    public int depthSumInverse(List<NestedInteger> nestedList) {
        Queue<NestedInteger> queue = new LinkedList<>();

        for(NestedInteger nestedInteger: nestedList) {
            if(nestedInteger != null) queue.offer(nestedInteger);
        }

        int res = 0;
        int prevSum = 0;
        while(!queue.isEmpty()) {
            int levelSum = 0;
            int size = queue.size();
            while(size > 0) {
                NestedInteger nestedInteger = queue.poll();
                if(nestedInteger.isInteger()) {
                    levelSum += nestedInteger.getInteger();
                }
                else {
                    for(NestedInteger subNestedInteger: nestedInteger.getList()) {
                        if(subNestedInteger != null) queue.offer(subNestedInteger);
                    }
                }
                size--;
            }
            res += prevSum + levelSum;
            prevSum += levelSum;
        }
        return res;
    }
}
